File: DSA/Navneet_590012555/Day13/Q2.java

String betterString(String s1, String s2) {
    return count(s1) >= count(s2) ? s1 : s2;
}

static int count(String s) {
    int n = s.length();
    int[] dp = new int[n + 1];
    int[] last = new int[256];

    dp[0] = 1;

    for (int i = 0; i < 256; i++)
        last[i] = -1;

    for (int i = 1; i <= n; i++) {
        dp[i] = 2 * dp[i - 1];

        char c = s.charAt(i - 1);

        if (last[c] != -1)
            dp[i] -= dp[last[c] - 1];

        last[c] = i;
    }

    return dp[n];
}

File: DSA/Khushi_590028673/Day13/Q2.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

long long countDistinctSubsequences(string s)
{
    int n = s.length();

    vector<long long> dp(n + 1);

    unordered_map<char, int> last;

    dp[0] = 1;

    for (int i = 1; i <= n; i++)
    {
        dp[i] = 2 * dp[i - 1];

        char ch = s[i - 1];

        if (last.find(ch) != last.end())
        {
            dp[i] -= dp[last[ch] - 1];
        }

        last[ch] = i;
    }

    return dp[n];
}

string betterString(string s1, string s2)
{
    long long cnt1 = countDistinctSubsequences(s1);
    long long cnt2 = countDistinctSubsequences(s2);

    if (cnt1 >= cnt2)
        return s1;

    return s2;
}

int main()
{
    string s1 = "abc";
    string s2 = "aac";

    cout << betterString(s1, s2);

    return 0;
}
